File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListPresenter.java
package com.app.sid.funwithflags.view.countrylist;

import android.support.annotation.NonNull;

import com.app.sid.funwithflags.datasets.remote.Country;

import java.util.List;

import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by Sidhanshu Udawat on 19-Jan-18.
 */

public class CountryListPresenter {

    @NonNull
    private final CountryListMvp.View mView;
    @NonNull
    private CompositeSubscription mSubscriptions;
    @NonNull
    private final CountryListMvp.Interactor mInteractor;

    public CountryListPresenter(@NonNull CountryListMvp.View view, CountryListMvp.Interactor interactor) {
        mView = checkNotNull(view, "view cannot be null!");
        mInteractor = interactor;
        mSubscriptions = new CompositeSubscription();
    }


    public void onPause() {
        unsubscribe();
    }

    public void init() {
        getCountries();
    }

    public void onRetry() {
        mView.showInternetError(true);
        getCountries();
    }

    public void getCountries() {
        if (!mView.isInternetAvailable()) {
            mView.showInternetError(true);
        } else {
            mView.showProgress(true);
            mSubscriptions.clear();

            Subscription subscription = mInteractor.getCountryList()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe(new Subscriber<List<Country>>() {
                        @Override
                        public void onCompleted() {
                        }

                        @Override
                        public void onError(Throwable e) {
                            loadingCountriesFailed();
                        }

                        @Override
                        public void onNext(List<Country> countryDTOS) {
                            loadCountries(countryDTOS);
                        }
                    });

            mSubscriptions.add(subscription);
        }
    }

    private void loadingCountriesFailed() {
        mView.showProgress(false);
        mView.showInternetError(true);
    }

    private void loadCountries(List<Country> countries) {
        mView.updateCountries(countries);
        mView.showProgress(false);
    }

    public void onQueryTextSubmit(String query) {
        mView.onQueryTextSubmit(query);
    }

    public void onQueryTextChange(String query) {
        mView.onQueryTextChange(query);
    }

    public void unsubscribe() {
        mSubscriptions.clear();
    }
}

File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListPresenter.kt
package com.app.sid.funwithflags.view.countrylist

import com.app.sid.funwithflags.datasets.remote.Country
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

class CountryListPresenter @Inject constructor(
        private val interactor: CountryListMvp.Interactor
) {
    private lateinit var view: CountryListMvp.View
    private val subscriptions: CompositeSubscription = CompositeSubscription()

    fun onPause() {
        unsubscribe()
    }

    fun bind(view: CountryListMvp.View) {
        this.view = view
        fetchCountries()
    }

    fun onRetry() {
        view.showInternetError(true)
        fetchCountries()
    }

    private fun fetchCountries() {
        if (!view.isInternetAvailable()) {
            view.showInternetError(true)
        } else {
            view.showProgress(true)
            subscriptions.clear()
            val subscription = interactor.getCountryList()
                    ?.observeOn(AndroidSchedulers.mainThread())
                    ?.subscribeOn(Schedulers.io())
                    ?.subscribe(object : Subscriber<List<Country?>?>() {
                        override fun onCompleted() {}
                        override fun onError(e: Throwable) {
                            loadingCountriesFailed()
                        }

                        override fun onNext(countries: List<Country?>?) {
                            countries?.filterNotNull()?.let {
                                loadCountries(it)
                            }
                        }
                    })
            subscriptions.add(subscription)
        }
    }

    private fun loadingCountriesFailed() {
        view.showProgress(false)
        view.showInternetError(true)
    }

    private fun loadCountries(countries: List<Country>) {
        view.updateCountries(countries)
        view.showProgress(false)
    }

    fun onQueryTextSubmit(query: String) {
        view.onQueryTextSubmit(query)
    }

    fun onQueryTextChange(query: String) {
        view.onQueryTextChange(query)
    }

    private fun unsubscribe() {
        subscriptions.clear()
    }

}